README.md
# elmo-diff

[![Build Status](https://travis-ci.org/geezee/elmo-diff.svg?branch=master)](https://travis-ci.org/geezee/elmo-diff)
[![Coverage Status](https://coveralls.io/repos/github/geezee/elmo-diff/badge.svg)](https://coveralls.io/github/geezee/elmo-diff)

Originally [elm-diff](../../../elm-diff), the library has been rewritten in ECMAScript 6
due to performance issues. The algorithm is inspired from Myers' diffing algorithm
that can be found [here](https://neil.fraser.name/writing/diff/myers.pdf), however
it is modified to make it probabilistic and quicker.

This library is not meant to visualize diffs between two strings. Its job is to
produce a diff from two strings and reconstruct a string from a source and a diff.

## Example

```js
// Finding the difference between two strings can be done in two ways

// detailed way:
const diff = new Diff("ABCABBA", "CBABAC");
const path = diff.computePath();
const serialized = diff.serialize(path);
console.log(serialized); // "0,1,1:B,5,5:C"

// or using the wrapper
console.log(Diff.diff("ABCABBA", "CBABAC")); // "0,1,1:B,5,5:C"

// to apply a diff to string
console.log(Diff.apply("ABCABBA", serialized)); // "CBABAC"
```

## Format

The format of the diff can be described as follows:

```
diff ::= <edit> | <edit> ',' <diff>
edit ::= <remove> | <insert>
remove ::= <index> | <index> '-' <index>
insert ::= <index> ':' <character>+

index ::= \d+
character ::= [^,\] | '\,' | '\\'
```

For example the diff `0,1:c,7,8:de,10-15` describes the operations:
* delete character at index 0
* insert character `c` at index 1
* delete character at index 7
* insert string `de` at index 8
* delete all characters from index `10` to `15`

## API

### `Diff#constructor(source, target)`

* `source` (string) the original string.
* `target` (string) the target string.

Refer to the documentation inside [src/diff.js](./src/diff.js) for a detailed
documentation of every non-static method of the class.

### `Diff#setSampleDecider(decider)`

* `decider` (function) must return a boolean.

**@return (this)**

Change the function that decides whether to sample or pop during the computation
of the path.

### `Diff#setPostStep(postStep)`

* `postStep` (function)

Set a function to be called after every step of the path computation.

### `static Diff.diff(source, target[, trials = 3])`

* `source` (string) the original string.
* `target` (string) the target string.
* `trials` (uint) the number of trials to do before returning the shortest path found (default 3).

**@return (string)**

Output a serialized diff path. Due to the probabilistic nature of the algorithm
the operation is done many times, the shortest diff is chosen to be returned.
By default 3 trials are done.

### `static Diff.apply(input, serializedPath)`

* `input` (string) the input to apply the diff to.
* `serializedPath` (string) a serialized diff produced by the module

**@return (string)**

Assuming the format of `serializedPath` is as expected, output the edit operations
described in the diff to the input.

(In theory) the following should always be correct:
```js
Diff.apply(str1, Diff.diff(str1, str2)) == str2
```

## A bit more about the algorithm

The algorithm traverses the edit graph much like how an A\* algorithm would find
the shortest path with the exception that popping is probabilistic. This entails
that the path found may not be the shortest one, however this allows to easily
explore potentially interesting parts of the edit graph.

The score function of a node `n` is:

    score(n) = max(n.x + n.y,  n.x + n.y - d(n, e)  - 2 * n.rootpathLength

where `d` is the l2 norm, `e` is the target to reach in the graph, and
`n.rootpathLength` is the length of the chain of parents to get from `n` to the root.

In every iteration of the algorithm there is a choice to either pick a sample
or to pop from the queue. The condition is:

    random() < t ^ -0.45

where `t` is the number of iteration

### Benchmarks

These numbers are based on 100 trials.

```
> npm run benchmark

Let t be the iteration number


function        time    length          min     max     range
1/t             4.5     1879.38         1835    1885    50
1/sqrt(t)       5.53    1868.72         1787    1897    110
1/log(t)        8.45    1864.28         1789    1905    116
t^-0.666        4.84    1872.22         1791    1891    100
t^-0.333        7.07    1859.26         1781    1899    118
t^-0.125        89.13   1889.6          1833    1933    100
exp(-t/2)       3.97    1881            1881    1881    0
```

With some interesting candidates with 1,000 trials

```
function        time    length          min     max     range
t^-0.666 [1k]   4.85    1874.4          1827    1891    64
t^-0.500 [1k]   4.97    1863.28         1817    1889    72
t^-0.450 [1k]   5.63    1865.1          1775    1895    120
t^-0.333 [1k]   7.09    1859.14         1781    1893    112
```


## License

See [LICENSE](./LICENSE)


CHANGELOG.md
# Changelog

## 1.0.0 - 02-04-2018

* Diff strings
* Apply diff to strings
* Control when to sample and when to pop from the queue

## 1.1.0 - 21-04-2018

* Consecutive inserts/deletes are merged to make a shorter diff: [Pull request](https://github.com/geezee/elmo-diff/pull/1)


test/dev.js
const Diff = require('../src/diff.js');

const source = 'm3lr2xczfuo1ljggqt8ct7';
const target = 'cm3lrv2czfuo1ljggqt8ct7';


let trg = ':-,\\ABCABBA THIS SHOULD BE DISREGARDED';
let src = 'CBABAC XXXX THIS SHOULD BE DISREGARDED XXXXX';

const diff = new Diff(src, trg);

const path = diff.computePath();
const serialized = diff.serialize(path);

const app = Diff.apply(src, serialized);

console.log('<< ', src);
console.log('>> ', trg);
console.log('%%%% ', serialized);
console.log('== ', app + '%');
console.log('?? ', app == trg);


test/test.js
const assert = require('assert');
const PriorityQueue = require('../src/queue.js');
const Diff = require('../src/diff.js');

describe('PriorityQueue', function() {
    describe('#size()', function() {
        it('size of fresh object is 0', function() {
            assert.equal(new PriorityQueue().size, 0);
        });
        it('size after push is 1', function() {
            assert.equal(new PriorityQueue().push(0, 1).size, 1);
        });
        it('size after push and pop is 0', function() {
            const q = new PriorityQueue();
            q.push(0, 1).pop();
            assert.equal(q.size, 0);
        });
    });


    describe('#isEmpty()', function() {
        it('true for fresh object', function() {
            assert.ok(new PriorityQueue().isEmpty());
        });
        it('false after push', function() {
            assert.ok(! new PriorityQueue().push(1, 0).isEmpty());
        });
        it('true after push and pop', function() {
            const q = new PriorityQueue();
            q.push(0, 1).pop();
            assert.ok(q.isEmpty());
        });
    });


    describe('#pop() and #peep()', function() {
        it('pop/peep empty queue is undefined', function() {
            assert.equal(new PriorityQueue().peep(), undefined);
            assert.equal(new PriorityQueue().pop(), undefined);
        });
        it('pop/peep queue with one element is the inserted element', function() {
            const q = new PriorityQueue();
            q.push(1, 0);
            assert.equal(q.peep().item, 1);
            assert.equal(q.pop().item, 1);
        });
        it('pop/peep queue with two elements returns the one with largest priority', function() {
            const q1 = new PriorityQueue();
            const q2 = new PriorityQueue();
            const bs = Math.random()*200;
            q1.push('a', bs); q1.push('b', bs+1);
            q2.push('b', bs+1); q2.push('b', bs);
            assert.equal(q1.peep().item, 'b');
            assert.equal(q2.peep().item, 'b');
            assert.equal(q1.pop().item, 'b');
            assert.equal(q2.pop().item, 'b');
        });
        it('after pop the queue property is still preserved', function() {
            const q = new PriorityQueue();
            q.push(2, 2).push(11, 11).push(3, 3).push(0, 0).push(5, 5).push(4, 4)
             .push(6, 6).push(7, 7).push(8, 8).push(9, 9).push(10, 10).push(1, 1)
             .push(12, 12).push(13, 13);
            q.pop();

            for (var i=0;2*i+1<q.size;i++) {
                assert.ok(q.items[i].score > q.items[2*i+1].score);
                if (2*i+2 < q.size) assert.ok(q.items[i].score > q.items[2*i+2].score);
            }
        });
    });

    describe('#remove(index)', function() {
        it('should preserve queue property after remove', function() {
            for (var j=1;j<14;j++) {
                const q = new PriorityQueue();
                q.push(2, 2).push(11, 11).push(3, 3).push(0, 0).push(5, 5).push(4, 4)
                 .push(6, 6).push(7, 7).push(8, 8).push(9, 9).push(10, 10).push(1, 1)
                 .push(12, 12).push(13, 13);
                q.remove(i);
                for (var i=0;2*i+1<q.size;i++) {
                    assert.ok(q.items[i].score > q.items[2*i+1].score);
                    if (2*i+2 < q.size) assert.ok(q.items[i].score > q.items[2*i+2].score);
                }
            }
        });
    });


    describe('#sample()', function() {
        it('sampling empty queue results in undefined', function() {
            assert.equal(new PriorityQueue().sample(), undefined);
        });
        it('sampling does actually sample', function() {
            const q = new PriorityQueue();
            const N = 1000;
            const probabilities = [1/6, 1/3, 1/2]
            const error = 0.05;

            q.push(0, 0); q.push(1, 1); q.push(2, 2);

            const counter = [0, 0, 0];
            for (var i=0;i<N;i++) {
                counter[q.sample().item]++;
            }
            
            assert.ok(Math.abs(counter[0]/N - 1/6) < error);
            assert.ok(Math.abs(counter[1]/N - 1/3) < error);
            assert.ok(Math.abs(counter[2]/N - 1/2) < error);
        });
    });


    describe('#push(item, score)', function() {
        it('queue property holds', function() {
            const q = new PriorityQueue();
            q.push(2, 2).push(11, 11).push(3, 3).push(0, 0).push(5, 5).push(4, 4)
             .push(6, 6).push(7, 7).push(8, 8).push(9, 9).push(10, 10).push(1, 1)
             .push(12, 12).push(13, 13);

            for (var i=0;2*i+1<q.size;i++) {
                assert.ok(q.items[i].score > q.items[2*i+1].score);
                if (2*i+2 < q.size) assert.ok(q.items[i].score > q.items[2*i+2].score);
            }
        });
    });


    describe('misc', function() {
        it('size after a random number of inserts is as expected', function() {
            const q = new PriorityQueue();
            const N = Math.ceil(100 * Math.random());
            for(var i=0;i<N;i++)
                q.push(i, Math.random()*200);
            assert.equal(q.size, N);
        });
        it('size after a random number of inserts/pops is as expected', function() {
            const q = new PriorityQueue();
            const N = Math.ceil(100 * Math.random());
            let inserts = 0; let deletes = 0;
            for(var i=0;i<N;i++) {
                if (Math.random() > 0.5) {
                    q.push(i, Math.random()*200);
                    inserts++;
                } else {
                    if (inserts - deletes > 0) deletes++;
                    q.pop();
                }
            }
            assert.equal(q.size, inserts - deletes);
        });
        it('pushing n elements and popping n will return a (desc) sorted list', function() {
            const N = Math.ceil(200 * Math.random() + 1);
            const q = new PriorityQueue();
            const arr = [];

            for(var i=0;i<N;i++) arr.push(i);
            for(var i=0;i<N/2;i++) {
                const j = Math.floor(Math.random() * N);
                const _tmp = arr[i]; arr[i] = arr[j]; arr[j] = _tmp;
            }

            q.toString;
            arr.forEach(elm => q.push(elm, elm));

            assert.equal(q.size, N);

            var i = N;
            while (!q.isEmpty()) {
                i--;
                assert.equal(q.pop().item, i);
            }
        });
    });
});


describe('Diff', function() {
    it('Myers example: serialization test', function() {
        const diff = new Diff("ABCABBA", "CBABAC");
        const path = [ {x: 7, y: 6}, {x: 7, y: 5}, {x: 5, y: 4}, {x: 3, y: 1},
            {x: 1, y: 0}, {x: 0, y: 0} ];

        const expected = '0,1,1:B,5,5:C';
        assert.equal(diff.serialize(path), expected);
    });
    it('Myers example: reconstruct from diff', function() {
        const diff = new Diff("ABCABBA", "CBABAC");
        assert.equal(Diff.apply("ABCABBA", '0,1,1:B,5,5:C'), "CBABAC");
        assert.equal(Diff.apply("ABCABBA", '0-1,1:B,5,5:C'), "CBABAC");
    });
    it('Not problematic when strings start with same letter', function() {
        const source = '6759o8ae24yduycft7nw';
        const target = '6759oae24yduycft7nw';
        const diff = new Diff(source, target);
        const path = diff.computePath();
        assert.equal(Diff.apply(source, diff.serialize(path)), target);
    });
    it('Not problematic when need to add', function() {
        const source = 'm3lr2xczfuo1ljggqt8ct7';
        const target = 'cm3lrv2czfuo1ljggqt8ct7';
        const diff = new Diff(source, target);
        const path = diff.serialize(diff.computePath());

        assert.equal(Diff.apply(source, path), target);
    });
    it('Not problematic if both strings are equal', function() {
        const str = 'hello world';
        assert.equal(Diff.diff(str,str), "");
        assert.equal(Diff.apply(str, ""), str);
    });
    it('Escapes commas and slashes', function() {
        assert.equal(Diff.diff('test', ',test'), '0:\\,');
        assert.equal(Diff.diff('test', '\\test'), '0:\\\\');
    });
    it('reconstructing a diffed pair works', function() {
        function randomDerivative(strSource) {
            const Ndiff = Math.ceil((0.05*Math.random()+0.05)*strSource.length);

            let randDiff = "";
            let index = 0;
            for (var i=0;i<Ndiff;i++) {
                if (Math.random() < 0.5) {
                    index += Math.floor(Math.random() * strSource.length / Ndiff);
                    randDiff += ',' + index;
                    if (Math.random() < 0.5) {
                        index += 1 + Math.floor(Math.random() * 5);
                        randDiff += '-' + index;
                    }
                } else {
                    let randc = String.fromCharCode(97 + Math.floor(Math.random()*26));
                    randDiff += ',' + index + ':' + randc;
                }
            }

            return Diff.apply(strSource, randDiff.substring(1));
        }

        for(var _=0;_<100;_++) {
            const N = Math.ceil(500 * Math.random()) + 50;
            const strSource = new Array(N).fill(0)
                .reduce(str => str + Math.random().toString(36).substring(2),"");
            const strTarget = randomDerivative(strSource);
            
            const diff = Diff.diff(strSource, strTarget, 1);
            assert.equal(Diff.apply(strSource, diff), strTarget);
        }
    });
});


src/diff.js
const PriorityQueue = require('./queue.js');

/**
 * A quick diffing algorithm inspired by [Mye86]. This implementation does not
 * always provide the smallest diff, or the "best". The algorithm can find a
 * diff that has a small (edit) length.
 *
 * @author george
 *
 * @example
 * ```
 * // to setup the diff algorithm
 * const diff = new Diff("ABCABBA", "CBABAC");
 *
 * // compute a short edit path
 * const path = diff.computePath();
 *
 * // diffing representation
 * const repr = diff.serialize(path);
 *
 * // or you can do it directly like this
 * const repr2 = Diff.diff("ABCABBA", "CBABAC");
 *
 * // transform input to output using the representation
 * const output = Diff.apply("ABCABBA", repr);
 *
 * assert(output === "CBABAC");
 * ```
 *
 * [Mye86] - https://neil.fraser.name/writing/diff/myers.pdf
 */
module.exports = class Diff {

    /**
     * Initialize the environment for the diffing.
     *
     * @param source    {string} the original string
     * @param target    {string} the target string
     */
    constructor(source, target) {
        if (source == "") throw "Source cannot be empty";
        if (target == "") throw "Target cannot be empty";

        this.source = source;
        this.target = target;

        this.iteration = 0;

        this.queue = new PriorityQueue();
        this.queue.push(this.newPoint(0, 0), 1);

        this.nodes = {};

        this.endpoint = this.newPoint(this.source.length, this.target.length);

        this.sampleDecider = () =>
            Math.random() < Math.pow(this.iteration, -0.45);

        this.postStep = function() { };
    }

    /**
     * Change the function that decides whether to sample or pop during the
     * computation of the path.
     *
     * @param decider {function : _ -> Bool}
     *
     * @return {Diff}
     */
    setSampleDecider(decider) {
        if (typeof decider == "function")
            this.sampleDecider = decider;
        return this;
    }

    /**
     * Set a function to be called after every step of the path computation.
     *
     * @param postStep {function}
     *
     * @return {Diff}
     */
    setPostStep(postStep) {
        if (typeof postStep == "function")
            this.postStep = postStep;
        return this;
    }

    newPoint(x, y) {
        return {
            x: x,
            y: y,
            id: x * this.source.length + y,
            dlength: 0,
            parent: null
        };
    }

    /**
     * Perform a step in computing a path
     */
    step() {
        this.iteration++;

        const shouldSample = this.sampleDecider();
        const sampleObj = shouldSample ? this.queue.sample() : this.queue.pop();
        const sample = sampleObj.item;

        if (this.isTarget(sample))
            return true;

        this.queue.remove(sampleObj.index);
        this.getNeighbors(sample).forEach(point => {
            const {x, y, dlength} = point;
            const de = Math.hypot(x-this.endpoint.x, y-this.endpoint.y);
            const score = Math.max(x+y, x + y - de) - 2 * dlength;

            point.dlength = sample.dlength + 1;

            if (point.id in this.nodes && this.nodes[point.id].dlength > point.dlength) {
                this.nodes[point.id].parent = sample.id;
                this.nodes[point.id].dlength = point.dlength;
            } else {
                point.parent = sample.id;
                this.queue.push(point, score);
                this.nodes[point.id] = point;
            }
        });

        this.postStep();

        return false;
    }

    /**
     * Compute the possible neighbors in the edit graph. If it's possible to
     * go along a diagonal, i.e. there is no changes to be done then the diagonal
     * point is the only neighbor. Otherwise the node to the right/bottom of the
     * node are the neighbors
     *
     * @param node    {Object.{x: uint, y: uint}} the node to compute its neighbors
     *
     * @return {List.{x: uint, y: uint}}
     */
    getNeighbors({x, y}) {
        const followDiagonal = ({x, y}) => {
            while (x < this.source.length && y < this.target.length
                && this.source.charAt(x) == this.target.charAt(y)) {
                x++;
                y++;
            }
            return this.newPoint(x, y);
        }

        const neighbors = [];
        const hasDiagonal = this.source.charAt(x) == this.target.charAt(y);

        if (x == this.endpoint.x && y == this.endpoint.y)
            return [ this.endpoint ];
        if (!hasDiagonal) {
            if (x + 1 <= this.source.length)
                neighbors.push(followDiagonal({ x: x+1, y: y }));
            if (y + 1 <= this.target.length)
                neighbors.push(followDiagonal({ x: x, y: y+1 }));
        } else neighbors.push(followDiagonal({ x: x+1, y: y+1 }));

        return neighbors
    }

    /**
     * Check if the point in question is the target node in the edit graph,
     * i.e. the bottom-most right-most node in the edit graph
     *
     * @return {boolean}
     */
    isTarget({x, y}) {
        return x == this.source.length
            && y == this.target.length;
    }

    /**
     * Perform iteration steps till the endpoint is reached, then compute a
     * path from the endpoint to the start and return it. The list starts with
     * the endpoint and ends with the starting point. In the list any node at
     * index i will have its parent at index i+1.
     *
     * @return {List.{x: uint, y: uint}}
     */
    computePath() {
        while(!this.step());

        const path = [];

        let current = this.endpoint.id;
        do {
            path.push(this.nodes[current]);
            current = this.nodes[current].parent;
        } while (current != null && current != 0);

        path.push(this.newPoint(0, 0));

        return path;
    }

    /**
     * Draw the edit graph with visited nodes highlighted
     *
     * @return {string}
     */
    visualize() {
        var str = "";
        for (var x=0;x<this.source.length;x++) {
            for (var y=0;y<this.target.length;y++)
                str += (x * this.source.length + y in this.nodes) ? 'o' : '-';
            str += "\n";
        }
        return str;
    }

    /**
     * Produce a representation of a path according to the format described
     * in the README file. Usually the input to this function is a result
     * of calling Diff#computePath()
     *
     * @param path  {List.{x: uint, y: uint}} the path to serialize
     *
     * @return {string}
     */
    serialize(path) {
        /* A transition is an insert (along y) or a delete (along x), it's also
         * followed by a tail; a long diagonal. To find the transition type
         * between (x1, y1) and (x2, y2) we need to unwind the diagonal and
         * then we can decide
         */
        function transitionType(src, trgt) {
            let source = { x: src.x, y: src.y },
                target = { x: trgt.x, y: trgt.y };

            if (source.x == target.x && source.y == target.y)
                throw "No transition can be done";

            let m = Math.min( Math.abs(source.x - target.x),
                              Math.abs(source.y - target.y) );
            target.x -= m;
            target.y -= m;

            if (source.x == target.x && source.y + 1 == target.y) return "i";
            else if (source.x + 1 == target.x && source.y == target.y) return "d";
        }

        function char_escape(chr) {
            switch (chr) {
                case ',': return '\\,';
                case '\\': return '\\\\';
                default: return chr;
            }
        }

        path.reverse();

        let result = "";
        for(var i=0;i<path.length-1;i++) {
            const par = i - 1 > 0 ? path[i-1] : null;
            const node = path[i];
            const succ = i + 1 < path.length - 1 ? path[i+1] : null;

            const type = transitionType(path[i], path[i+1]);

            if (succ != null) succ.parent_type = type;
            if (node != null) node.type = type;
            if (succ !== null) succ.type = transitionType(path[i+1], path[i+2]);

            // Sorry for the ugly syntax, i'll explain:
            if (type == 'i') {
                /* It's not efficient to output `4:A,5:B,6:C`, it's better to store
                 * it as `4:ABC`. This merger of inserts can be decided in the second,
                 * third, etc... insert. It's not only sufficient to test if the parent
                 * is also an insert, consider `0:A,1:B,3:C` it's not correct to merge
                 * it as `0:ABC`, the correct way would be `0:AB,3:C`. So the current
                 * insert index must be the successor of the parent's
                 */
                let c = char_escape(this.target.charAt(node.y));
                let validPar = typeof par === 'object' && par !== null;

                result += node.parent_type == 'i' && validPar && par.y + 1 == node.y
                    ? c
                    : ',' + node.y + ':' + c;

            } else if (type == 'd') {
                /* It's also not efficient to output `7,8,9,10,11,12,13,14` instead
                 * we can output `7-14`
                 *
                 * The start of a range satisfies one of the following:
                 * i.   start of the path
                 * ii.  parent is an insert
                 * iii. parent doesn't delete at the previous index
                 *
                 * The end of a range satisfies one of the following:
                 * i.   end of the path
                 * ii.  successor is an insert
                 * iii. successor doesn't delete at the next index
                 *
                 * Singletons satisfy both start and end of path.
                 */
                const start = (par === null)
                           || (par !== null &&
                                   ((par.type == 'i')
                                 || (par.type == 'd' && par.x + 1 < node.x)));

                const end = (succ === null)
                         || (succ !== null &&
                                  ((succ.type == 'i')
                                || (succ.type == 'd' && succ.x - 1 > node.x)));

                if (start) result += ',' + node.x; // singletons and start of ranges
                if (!start && end) result += '-' + node.x; // end of ranges
            }
        }

        return result.substr(1);
    }

    /**
     * Assuming the format of `serializedPath` is as expected, output the edit
     * operations described in the diff to the input.
     *
     * @param input {string} the input to apply the diff to.
     * @param serializedPath {string} a serialized diff produced by the module
     *
     * @return {string}
     */
    static apply(input, serializedPath) {
        let delOffset = 0;
        let result = input;

        function apply_op(op) {
            let matches;
            if ((matches = /^(\d+):(.*)$/g.exec(op)) !== null) {
                let i = parseInt(matches[1]);
                result = result.substring(0, i) + matches[2] + result.substring(i);
                delOffset -= matches[2].length;
            } else if ((matches = /^(\d+)-(\d+)$/g.exec(op)) !== null) {
                let start = parseInt(matches[1]) - delOffset,
                    end = parseInt(matches[2]) - delOffset;
                result = result.substring(0, start) + result.substring(end+1);
                delOffset += end - start + 1;
            } else if ((matches = /^(\d+)$/g.exec(op)) !== null) {
                let i = parseInt(op) - delOffset;
                result = result.substring(0, i) + result.substring(i+1);
                delOffset++;
            } else {
                throw "Operation not valid: " + op;
            }
        }

        let op = '';
        let escaped = false;

        for (var i=0;i<serializedPath.length;i++) {
            if (serializedPath.charAt(i) == ',' && !escaped) {
                apply_op(op);
                op = '';
            } else if (serializedPath.charAt(i) == '\\' && !escaped) {
                escaped = true;
                continue;
            } else {
                op += serializedPath.charAt(i);
                escaped = false;
            }
        }

        if (op.length > 0) {
            apply_op(op);
        }

        return result;
    }

    /**
     * Output a serialized diff path. Due to the probabilistic nature of the
     * algorithm the operation is done many times, the shortest diff is chosen
     * to be returned. By default 3 trials are done.
     *
     * @param source    {string} the original string
     * @param target    {string} he target string
     * @param trials    {uint} the number of trials to do before returning the
     *                  shortest path found (default 3)
     *
     * @return {string}
     */
    static diff(source, target, trials = 3) {
        let shortest = null;
        for(var _=0;_<trials;_++) {
            const diff = new Diff(source, target);
            const path = diff.computePath();
            if (shortest == null || path.length < shortest.length)
                shortest = path;
        }
        return new Diff(source, target).serialize(shortest);
    }

}


benchmarks/benchmarks.js
const Diff = require('../src/diff.js');

const str1 = "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Vi" +
    "vamus eget interdum enim, sit amet placerat justo. Donec suscipit fe" +
    "ugiat nunc non luctus. Etiam at ante auctor, viverra urna ut, effici" +
    "tur justo. Proin tincidunt commodo sapien, ut blandit augue placerat" +
    "id. Nunc lobortis justo sit amet leo interdum, non auctor elit biben" +
    "dum. In hac habitasse platea dictumst. Aenean varius fermentum diam " +
    "semper gravida. Maecenas ultrices venenatis leo, non porta diam vive" +
    "rra et. Maecenas in ante faucibus, blandit tellus ac, venenatis odio" +
    ". Duis massa quam, interdum id elit eget, ultricies dictum enim. Sed" +
    "dictum pharetra ipsum, eget placerat tortor sodales eget. Quisque en" +
    "im sapien, cursus ut eleifend ut, blandit finibus mauris.  Aliquam f" +
    "inibus venenatis lorem in vehicula. Fusce arcu enim, cursus in lorem" +
    "a, semper faucibus lacus. Nullam rhoncus felis sed egestas vulputate" +
    ". Sed sapien leo, hendrerit vel nibh sed, commodo porttitor mi. Null" +
    "a nec sollicitudin lacus, non tristique sapien.";

const str2 = "Proin efficitur eros eu varius varius. Aenean quis diam ac " +
    "orci porttitor vulputate. Cras vitae fringilla libero, at dictum nun" +
    "c. Sed rutrum eu orci eu bibendum. Nam pharetra vulputate lacus, in " +
    "cursus libero porta nec. Aenean finibus arcu ex, ac scelerisque nequ" +
    "e convallis vel. Nullam eu suscipit nisi. Pellentesque rhoncus portt" +
    "itor nulla interdum mattis. Aliquam et sagittis dolor, eu volutpat f" +
    "elis. Ut vitae sem arcu. Praesent sed sem ac leo auctor feugiat. Don" +
    "ec dapibus arcu velit, vehicula pretium dui viverra in. Suspendisse " +
    "tristique suscipit congue.  Maecenas ultrices venenatis leo, non por" +
    "ta diam viverra et. Maecenas in ante faucibus, blandit tellus ac, ve" +
    "nenatis odio. Duis massa quam, interdum id elit eget, ultricies dict" +
    "um enim. Sed dictum pharetra ipsum, eget placerat tortor sodales ege" +
    "t. Quisque enim sapien, cursus ut eleifend ut, blandit finibus mauri" +
    "s. Aliquam finibus VENEnatis lorem In Vehicula. Fusce arcu ENIM, CUR" +
    "sus eth in lorem ar faucibus lacus. nULLAM rhonfelis sed gs th hta  " +
    "vulPuTatEsed sapien leO,  HENdrerel NIBH SED, Commodo portloe la nec" +
    "sollics, nonRISTIque sapien. ";

const benchmark = function(name, code, N=100) {
    const round = a => Math.round(a * 100) / 100;
    let total = 0;
    let length = 0;
    let min = str1.length + str2.length;
    let max = 0;
    for(var i=0;i<N;i++) {
        const start = new Date().getTime();
        const value = code();
        const end = new Date().getTime();

        total += (end - start) / N;
        length += value / N;
        min = Math.min(min, value);
        max = Math.max(max, value);
    }
    console.log(name+''+round(total)+'\t'+round(length)+'\t\t'+min+'\t'+max+'\t'+(max-min));
}

const makeSuite = function(name, sampleDecider, postStep) {
    var namepad = "                "; // 16 whitespaces
    name += namepad.substring(0, 16 - name.length);
    name = name.substring(0, 16);

    benchmark(name, function() {
        let n = 0, t = 0;
        const diff = new Diff(str1, str2);

        diff.setPostStep((typeof postStep == "function") ?
            (() => { t++; postStep(); }) : (() => t++));

        diff.setSampleDecider(() => Math.random () < sampleDecider(n++, t));

        return diff.computePath().length;
    });
};

console.log('Let t be the iteration number\n\n');

console.log('function\ttime\tlength\t\tmin\tmax\trange');

makeSuite('1/t', (n, t) => 1 / t);
makeSuite('1/sqrt(t)', (n, t) => 1 / Math.sqrt(t));
makeSuite('1/log(t)', (n, t) => 1 / Math.log(t));
makeSuite('t^-0.666', (n, t) => Math.pow(t, -2.0/3));
makeSuite('t^-0.333', (n, t) => Math.pow(t, -1.0/3));
makeSuite('t^-0.125', (n, t) => Math.pow(t, -0.125));
makeSuite('exp(-t/2)', (n, t) => Math.exp(-t/2));

console.log('');

makeSuite('t^-0.666 [1k]', (n, t) => Math.pow(t, -2.0/3), 1000);
makeSuite('t^-0.500 [1k]', (n, t) => 1 / Math.sqrt(t), 1000);
makeSuite('t^-0.450 [1k]', (n, t) => Math.pow(t, -0.45), 1000);
makeSuite('t^-0.333 [1k]', (n, t) => Math.pow(t, -1.0/3), 1000);
